tracer/src/Datadog.Trace.SourceGenerators/AspectsDefinitions/AspectsDefinitionsGenerator.cs

```diff
@@ -21,181 +22,216 @@
-                }
+                sb.AppendLine(FormatLine(aspect));
             }
         }
 
-        return null;
+        sb.AppendLine("""
+        };
     }
+}
+""");
 
-    private static IReadOnlyList<string> GetDefinitionsToWrite(Compilation compilation, ImmutableArray<ClassDeclarationSyntax> classes, CancellationToken ct)
-    {
-        if (classes.IsDefaultOrEmpty)
+        context.AddSource("AspectsDefinitions.g.cs", SourceText.From(sb.ToString(), Encoding.UTF8));
+
+        string FormatLine(string line)
         {
-            // nothing to do yet
-            return Array.Empty<string>();
+            return $"\"{line.Replace("\"", "\\\"")}\",";
         }
-
-        return GetAspectSources(compilation, classes.Distinct(), ct);
     }
 
-    private static void Execute(IReadOnlyList<string> definitions, SourceProductionContext context)
+    private static Result<(ClassAspects Aspects, bool IsValid)> GetAspectsToGenerate(GeneratorAttributeSyntaxContext context, CancellationToken ct)
     {
-        string source = GetSource(definitions);
-        context.AddSource("AspectsDefinitions.g.cs", SourceText.From(source, Encoding.UTF8));
-    }
-
-    private static IReadOnlyList<string> GetAspectSources(Compilation compilation, IEnumerable<ClassDeclarationSyntax> classes, CancellationToken cancellationToken)
-    {
-        INamedTypeSymbol? aspectClassAttribute = compilation.GetTypeByMetadataName(Constants.AspectClassAttributeFullName);
-        if (aspectClassAttribute is null)
+        INamedTypeSymbol? classSymbol = context.TargetSymbol as INamedTypeSymbol;
+        if (classSymbol is null)
         {
             // nothing to do if this type isn't available
-            return Array.Empty<string>();
+            return new Result<(ClassAspects, bool)>((default, false), default);
         }
 
-        var results = new List<string>();
-
-        List<(string, INamedTypeSymbol)> classSymbols = new List<(string, INamedTypeSymbol)>();
+        ct.ThrowIfCancellationRequested();
+        List<DiagnosticInfo>? diagnostics = null;
 
-        foreach (ClassDeclarationSyntax classDec in classes)
+        string aspectClass = string.Empty;
+        foreach (AttributeData attribute in classSymbol.GetAttributes())
         {
-            // stop if we're asked to
-            cancellationToken.ThrowIfCancellationRequested();
-
-            SemanticModel sm = compilation.GetSemanticModel(classDec.SyntaxTree);
-            INamedTypeSymbol? classSymbol = sm.GetDeclaredSymbol(classDec, cancellationToken) as INamedTypeSymbol;
-            Debug.Assert(classSymbol is not null, "Instrumented class is not present");
-            var boundAttributes = classSymbol!.GetAttributes();
-
-            // Process InstrumentMethodAttribute first
-            foreach (AttributeData attributeData in boundAttributes)
+            if (attribute.AttributeClass is null) { continue; }
+            if (attribute.AttributeClass.Name == "AspectClassAttribute")
```

Comment:
You _may_ want to use the full type name to be safe (though in reality, it almost certainly won't be a problem!)

https://github.com/DataDog/dd-trace-dotnet/blob/40a49fb63684c6ae38224ffc2d83e7f233da6e0f/tracer/src/Datadog.Trace.SourceGenerators/EnumExtensions/EnumExtensionsGenerator.cs#L76-L78